components/eamxx/src/share/field/field_utils_impl.hpp

```diff
@@ -322,31 +322,75 @@ void horiz_contraction(const Field &f_out, const Field &f_in,
 
   const int ncols = l_in.dim(0);
 
+  bool is_masked = f_in.get_header().has_extra_data("mask_data");
+  bool is_avg_masked = AVG && is_masked;
+  bool is_comm_avg_masked = comm && is_avg_masked;
+
+  if (is_comm_avg_masked) {
+    // make sure f_tmp is allocated correctly
+    EKAT_REQUIRE_MSG(f_tmp.data_type() == f_in.data_type(),
+                     "Error! Temporary field must have the same data type as input fields.");
+    EKAT_REQUIRE_MSG(f_tmp.is_allocated(),
+                     "Error! Temporary field must be allocated.");
+    EKAT_REQUIRE_MSG(f_tmp.get_header().get_identifier().get_layout() == l_out,
+                     "Error! Temporary field must have the same layout as output field.");
+  }
+
   switch(l_in.rank()) {
     case 1: {
       auto v_in  = f_in.get_view<const ST *>();
+      auto v_m   = is_masked ? f_in.get_header().get_extra_data<Field>("mask_data").get_view<const ST *>() : v_in;
       auto v_out = f_out.get_view<ST>();
+      auto v_tmp = is_comm_avg_masked ? f_tmp.get_view<ST>() : v_out; 
+      ST n = 0, d = 0;
       Kokkos::parallel_reduce(
           f_out.name(), RangePolicy(0, ncols),
-          KOKKOS_LAMBDA(const int i, ST &ls) { ls += v_w(i) * v_in(i); },
-          v_out);
+          KOKKOS_LAMBDA(const int i, ST &n_acc, ST &d_acc) {
+            auto mask = is_masked ? v_m(i) : ST(1.0);
+            n_acc += v_w(i) * v_in(i) * mask;
+            d_acc += v_w(i) * mask;
+          },
+          Kokkos::Sum<ST>(n), Kokkos::Sum<ST>(d));
+      Kokkos::deep_copy(v_out, n);
```

Comment:
Looks like you should move this inside the first branch of the if, to avoid a pointless deep_copy in case the `else if` triggers?